compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/hotspot/aarch64/shenandoah/AArch64HotSpotShenandoahLoadRefBarrierOp.java

```diff
@@ -0,0 +1,233 @@
+            masm.mov(64, resultRegister, objectRegister);
+
+            if (!notNull) {
+                // Check for object being null.
+                masm.cbz(64, resultRegister, done);
+            }
+
+            // Check for heap stability
+            int gcStateOffset = HotSpotReplacementsUtil.shenandoahGCStateOffset(config);
+            AArch64Address gcState = masm.makeAddress(8, thread, gcStateOffset);
+            masm.ldr(8, rscratch1, gcState);
+            if (strength != ShenandoahLoadRefBarrierNode.ReferenceStrength.STRONG) {
+                // This is needed because in a short-cut cycle we may get a trailing
+                // weak-roots phase but no evacuation/update-refs phase, and during that,
+                // we need to take the LRB to report null for unreachable weak-refs.
+                // This is true even for non-cset objects.
+                // Two tests because HAS_FORWARDED | WEAK_ROOTS currently is not representable
+                // as a single immediate.
+                masm.tst(64, rscratch1, GCState.HAS_FORWARDED.value);
+                masm.branchConditionally(AArch64Assembler.ConditionFlag.NE, slowPath);
+                masm.tst(64, rscratch1, GCState.WEAK_ROOTS.value);
+                masm.branchConditionally(AArch64Assembler.ConditionFlag.NE, slowPath);
+            } else {
+                masm.tst(64, rscratch1, GCState.HAS_FORWARDED.value);
+                masm.branchConditionally(AArch64Assembler.ConditionFlag.NE, csetCheck);
+            }
+            masm.bind(done);
+
+            // Check for object in collection set in an out-of-line mid-path.
+            if (strength == ShenandoahLoadRefBarrierNode.ReferenceStrength.STRONG) {
+                crb.getLIR().addSlowPath(this, () -> {
+                    try (AArch64MacroAssembler.ScratchRegister tmp1 = masm.getScratchRegister(); AArch64MacroAssembler.ScratchRegister tmp2 = masm.getScratchRegister()) {
+                        Register rtmp1 = tmp1.getRegister();
+                        Register rtmp2 = tmp2.getRegister();
+                        masm.bind(csetCheck);
+                        masm.mov(rtmp1, HotSpotReplacementsUtil.shenandoahGCCSetFastTestAddr(config));
+                        masm.lsr(64, rtmp2, objectRegister, HotSpotReplacementsUtil.shenandoahGCRegionSizeBytesShift(config));
+                        masm.ldr(8, rtmp2, AArch64Address.createRegisterOffsetAddress(8, rtmp1, rtmp2, false));
+                        masm.cbnz(8, rtmp2, slowPath);
+                        masm.jmp(done);
+                    }
+                });
+            }
+            // Call runtime slow-path LRB in out-of-line slow-path.
+            crb.getLIR().addSlowPath(this, () -> {
+                try (AArch64MacroAssembler.ScratchRegister tmp1 = masm.getScratchRegister(); AArch64MacroAssembler.ScratchRegister tmp2 = masm.getScratchRegister()) {
```

Comment:
@tkrodriguez will we generate multiple copies of this in the same compilation unit?

Reply:
Yes, `addSlowPath` always makes a unique copy of the slow path since it's dispatched with branches.  Each copy is slightly distinct because of register usage.  For the barriers, we could make these more compact by having custom slow path stubs with entry points specialized by register signature but that would only save a few instructions but maybe it would add up.